run_tests.py

```diff
@@ -53,117 +59,185 @@ def modified_environment(**kwargs):
+    """
 
-    port = str(get_free_ports(1)[0])
-    lewis_launcher.port = port
-    ioc_launcher.port = port
     try:
-        ioc_launcher.macros = m.MACROS
+        iocs = test_module.IOCS
     except AttributeError:
-        ioc_launcher.macros = {}
+        raise AttributeError("Expected module '{}' to contain an IOCS attribute".format(test_module.__name__))
+
+    if len(iocs) < 1:
+        raise ValueError("Need at least one IOC to launch")
+
+    print("Testing module {} in {} mode.".format(test_module.__name__, "recsim" if recsim else "devsim"))
+
+    device_launchers = []
+    for ioc in iocs:
+
+        free_port = str(get_free_ports(1)[0])
+        macros = ioc["macros"]
+        macros['EMULATOR_PORT'] = free_port
+
+        ioc_launcher = IocLauncher(device=ioc["name"],
+                                   directory=ioc["directory"],
+                                   macros=macros,
+                                   use_rec_sim=recsim,
+                                   var_dir=var_dir,
+                                   port=free_port)
+
+        if "emulator" in ioc and not recsim:
+
+            emulator_name = ioc["emulator"]
+            try:
+                emulator_protocol = ioc["emulator_protocol"]
+            except KeyError:
+                emulator_protocol = "stream"
+
+            try:
+                emulator_device_package = ioc["emulator_package"]
+            except KeyError:
+                emulator_device_package = "lewis_emulators"
+
+            try:
+                emulator_path = ioc["emulator_path"]
+            except KeyError:
+                emulator_path = os.path.join(EPICS_TOP, "support", "DeviceEmulator", "master")
+
+            lewis_launcher = LewisLauncher(
+                device=emulator_name,
+                python_path=os.path.abspath(arguments.python_path),
+                lewis_path=os.path.abspath(arguments.emulator_path),
+                lewis_protocol=emulator_protocol,
+                lewis_additional_path=emulator_path,
+                lewis_package=emulator_device_package,
+                var_dir=var_dir,
+                port=free_port
+            )
+
+        elif "emulator" in ioc:
+            lewis_launcher = LewisNone(ioc["emulator"])
+        else:
+            lewis_launcher = None
+
+        device_launchers.append(device_launcher(ioc_launcher, lewis_launcher))
+
+    return device_launchers
+
+
+def load_module_by_name_and_run_tests(module_name):
+    test_module = load_module("tests.{}".format(module_name))
+
     try:
-        ioc_launcher.device_prefix = m.DEVICE_PREFIX
+        modes = test_module.TEST_MODES
     except AttributeError:
-        ioc_launcher.device_prefix = None
+        raise ValueError("Expected test module {} to contain a TEST_MODES attribute".format(test_module.__name__))
 
-    # Override any emulator port that might be set elsewhere
-    ioc_launcher.macros['EMULATOR_PORT'] = port
+    test_results = []
+    for mode in modes:
```

Comment:
This means that if a list of N elements then it will run the test in devsim N times, regardless of what those elements are. Maybe worth some checks that the list is a set and only contains valid TestModes?